File: ComputerVision/bresenham.py

def reflection():
    if -1 < m < 1:
        x, y = y, x
        switch_xy = True

    if x1 > x2:
        x1 = -x1
        x2 = -x2
        switch_x = True

    if y1 > y2:
        y1 = -y1
        y2 = -y2
        switch_y = True


def inv_reflecion():
    if switch_y:
        for point in points:
            point[1] = -points[1]

    if switch_x:
        for point in points:
            point[0] = -point[0]

    if switch_xy:
        for point in points:
            point = point[::-1]

def bresenham(p1, p2):
    # reflection (p1, p2)

    x1 = p1[0]
    y1 = p1[1]
    x2 = p2[0]
    y2 = p2[1]

    dx = p2[0] - p1[0]
    dy = p2[1] - p1[1]

    m = dy / dx
    e = (m - 1) /  2

    # draw (x1, y1)

    while x1 <= x2:
        if e >= 0:
            y1 += 1
            e -= 1
        x1 += 1
        e += m
        # draw (x1, y1)

    # inv_reflection()

File: ComputerVision/bresenham.cpp
#include <SFML/Graphics.hpp>
#include <functional>
#include <iostream>
#include <list>

using namespace std;

// Eq. to list of lists in python
typedef list <pair<int,int>> plist;

plist rasterBresenham(int x0, int y0, int x1, int y1)
{
	plist coords;
    int dx, dy, p, x, y;

	dx = x1 - x0;
	dy = y1 - y0;

	x = x0;
	y = y0;

	p =2 * dy - dx;

	// TODO: adapt to x > x1
	while (x <= x1)
	{
		coords.push_back(make_pair(x, y));
		/* cout << x << y << endl; */
		if (p >= 0)
		{
			y = y + 1;
			p = p + 2 * dy - 2 * dx;
		}
		else
		{
			p = p + 2 * dy;
		}
		x = x + 1;
	}

	return coords;
}


void printList(plist::reference x)
{
  cout << x.first << " " << x.second << endl;
}


int main()
{
	int x0, y0, x1, y1;

	cout<<"Enter co-ordinates of first point: ";
	cin >> x0 >> y0;

	cout<<"Enter co-ordinates of second point: ";
	cin >> x1 >> y1;

	plist coords = rasterBresenham(x0, y0, x1, y1);

	for_each(coords.begin(), coords.end(), ref(printList));

	int columns = 30;
    int rows = 30;
    sf::RenderWindow window(sf::VideoMode(825, 825), "Bresenham");
    sf::RectangleShape grid[columns][rows];


    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        window.clear();
        sf::Vector2f cellSize(22.5f, 22.5f);

		// Create Grid
        for(int i=0;i<columns;i++){
            for(int j=0;j<rows;j++){
                grid[i][j].setSize(cellSize);
                grid[i][j].setOutlineColor(sf::Color::Black);
                grid[i][j].setOutlineThickness(1.0f);

				// TODO: Change grid direction
                grid[i][j].setPosition(i*cellSize.x + 1.0f, j*cellSize.y + 1.0f);

                window.draw(grid[i][j]);

            }
        }

		// Plot Line
		for (auto const i : coords)
		{
			cout << i.first << i.second << endl;
			grid[i.first][i.second].setFillColor(sf::Color::Red);
			grid[i.first][i.second].setOutlineColor(sf::Color::Red);
			window.draw(grid[i.first][i.second]);
		}
        window.display();
    }

	return 0;
}
